gslb/ingestion/bootup_handler.go

```diff
@@ -0,0 +1,198 @@
+			// skip if no generation change, applicable to all objects
+			return e.ObjectOld.GetGeneration() != e.ObjectNew.GetGeneration()
+		},
+	}
+}
+
+func (r *AMKOClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	gslbutils.Debugf("Starting AMKOCluster reconcialation")
+
+	var amkoClusterList amkov1alpha1.AMKOClusterList
+	err := r.List(ctx, &amkoClusterList, &client.ListOptions{
+		Namespace: federator.AviSystemNS,
+	})
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("AMKOClusterObjects can't be listed, err: %v", err)
+	}
+	if len(amkoClusterList.Items) > 1 {
+		return reconcile.Result{}, fmt.Errorf("only one AMKOClusterObject allowed per cluster")
+	}
+
+	amkoCluster := amkoClusterList.Items[0]
+	if currentLeader != amkoCluster.Spec.IsLeader {
+		gslbutils.Warnf("AMKO leader flag changed, API server would be shut down")
+		apiserver.GetAmkoAPIServer().ShutDown()
+		return ctrl.Result{}, nil
+	}
+
+	currentLeader = amkoCluster.Spec.IsLeader
+	if !currentLeader {
+		gslbutils.Warnf("AMKO is not a leader, will return")
+		return ctrl.Result{}, nil
+	}
+
+	memberClusters, err := federator.FetchMemberClusterContexts(ctx, amkoCluster.DeepCopy())
+	gslbutils.Debugf("memberClusters obtained during reconcialation: %v", memberClusters)
+
+	err = federator.ValidateMemberClusters(ctx, memberClusters)
+	if err != nil {
+		apiserver.GetAmkoAPIServer().ShutDown()
+		return reconcile.Result{}, fmt.Errorf("error in validating the member clusters: %v", err)
+	}
+
+	gslbutils.Debugf("AMKOCluster reconcialation completed")
+	return ctrl.Result{}, nil
+}
+
+func (r *AMKOClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Watches(&source.Kind{Type: &amkov1alpha1.AMKOCluster{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) []reconcile.Request {
+				return []reconcile.Request{
+					{
+						NamespacedName: types.NamespacedName{
+							Name:      o.GetName(),
+							Namespace: o.GetNamespace(),
+						},
+					},
+				}
+			}),
+		).
+		For(&amkov1alpha1.AMKOCluster{}).
+		WithEventFilter(acceptGenerationChangePredicate()).
+		Complete(r)
+}
+
+func handleBootup(cfg *restclient.Config) (bool, error) {
+	clusterClient, _ := InitializeClusterClient(cfg)
+	var amkoClusterList amkov1alpha1.AMKOClusterList
+	clusterClient.List(context.TODO(), &amkoClusterList, &client.ListOptions{
+		Namespace: federator.AviSystemNS,
+	})
+
+	if len(amkoClusterList.Items) == 0 {
+		gslbutils.Logf("No AMKOCluster object found, AMKO would start as leader")
+		return true, nil
+	}
+	if len(amkoClusterList.Items) > 1 {
+		return false, fmt.Errorf("only one AMKOClusterObject allowed per cluster")
+	}
+
+	amkoCluster := amkoClusterList.Items[0]
+	if amkoCluster.Spec.IsLeader {
+		currentLeader = true
+		gslbutils.Logf("AMKOCluster object found and AMKO would start as leader")
+	} else {
+		gslbutils.Logf("AMKOCluster object found and AMKO would start as follower")
+	}
```

Comment:
We check for the number of items and `isLeader` flag above (in `Reconcile`) too. Can we put this into a function and call that instead?

Reply:
I think those checks are not identical and minimal. In Reconcile, we first check if the leader flag has changed.